File: packages/jvm/src/main/java/io/github/qorechain/rdk/util/Bytes.java
package io.github.qorechain.rdk.util;

import java.util.Base64;
import java.util.regex.Pattern;

/** Byte/hex helpers used across the tx, DA, and event layers. */
public final class Bytes {
    private Bytes() {}

    private static final char[] HEX = "0123456789abcdef".toCharArray();
    private static final Pattern HEX_RE = Pattern.compile("^[0-9a-fA-F]*$");
    private static final Pattern HEX_FULL_RE = Pattern.compile("^[0-9a-fA-F]+$");
    private static final Pattern BASE64ISH_RE = Pattern.compile("[+/=]|[g-zG-Z]");

    /** Convert bytes to a lowercase hex string (no {@code 0x} prefix). */
    public static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) {
            sb.append(HEX[(b >> 4) & 0xf]);
            sb.append(HEX[b & 0xf]);
        }
        return sb.toString();
    }

    /** Parse a hex string (with or without a {@code 0x} prefix) into bytes. */
    public static byte[] hexToBytes(String hex) {
        String h = hex;
        if (h.startsWith("0x") || h.startsWith("0X")) {
            h = h.substring(2);
        }
        if (h.length() % 2 != 0 || !HEX_RE.matcher(h).matches()) {
            throw new IllegalArgumentException("invalid hex string: \"" + hex + "\"");
        }
        byte[] out = new byte[h.length() / 2];
        for (int i = 0; i < out.length; i++) {
            out[i] = (byte) Integer.parseInt(h.substring(i * 2, i * 2 + 2), 16);
        }
        return out;
    }

    /** Coerce a hex string into bytes (string overload of the TS {@code toBytes} helper). */
    public static byte[] toBytes(String input) {
        return hexToBytes(input);
    }

    /** Decode a standard (RFC 4648) base64 string into bytes. */
    public static byte[] base64ToBytes(String b64) {
        if (b64.isEmpty()) {
            return new byte[0];
        }
        return Base64.getDecoder().decode(b64);
    }

    /**
     * Decode a bytes value as it arrives from the chain's wire surface. Cosmos gRPC-gateway
     * (jsonpb) encodes proto {@code bytes} as base64; some surfaces send hex. Disambiguated by
     * alphabet and length (a 32-byte root is 64 hex chars vs ~44 base64 chars with {@code +/=}),
     * so both encodings round-trip correctly.
     */
    public static byte[] decodeWireBytes(String value) {
        if (value == null || value.isEmpty()) {
            return new byte[0];
        }
        boolean isHex = value.length() % 2 == 0 && HEX_FULL_RE.matcher(value).matches();
        boolean looksBase64 = BASE64ISH_RE.matcher(value).find();
        return isHex && !looksBase64 ? hexToBytes(value) : base64ToBytes(value);
    }
}

File: packages/py/src/qorrdk/utils/bytes.py
"""Byte/hex helpers used across the tx, DA, and event layers."""

from __future__ import annotations

import base64
import re
from typing import Union

_HEX_RE = re.compile(r"^[0-9a-fA-F]*$")
_HEX_FULL_RE = re.compile(r"^[0-9a-fA-F]+$")
_BASE64_HINT_RE = re.compile(r"[+/=]|[g-zG-Z]")


def bytes_to_hex(data: bytes) -> str:
    """Convert bytes to a lowercase hex string (no ``0x`` prefix)."""
    return data.hex()


def hex_to_bytes(hex_str: str) -> bytes:
    """Parse a hex string (with or without a ``0x`` prefix) into bytes."""
    h = hex_str[2:] if hex_str[:2] in ("0x", "0X") else hex_str
    if len(h) % 2 != 0 or not _HEX_RE.match(h):
        raise ValueError(f'invalid hex string: "{hex_str}"')
    return bytes.fromhex(h)


def to_bytes(value: Union[str, bytes, bytearray]) -> bytes:
    """Coerce a hex string or byte sequence into bytes."""
    if isinstance(value, (bytes, bytearray)):
        return bytes(value)
    return hex_to_bytes(value)


def base64_to_bytes(b64: str) -> bytes:
    """Decode a standard (RFC 4648) base64 string into bytes."""
    if b64 == "":
        return b""
    return base64.b64decode(b64)


def bytes_to_base64(data: bytes) -> str:
    """Encode bytes as a standard (RFC 4648) base64 string."""
    return base64.b64encode(data).decode("ascii")


def decode_wire_bytes(value: str) -> bytes:
    """Decode a bytes value as it arrives from the chain's wire surface.

    Cosmos gRPC-gateway (jsonpb) encodes proto ``bytes`` as base64; some
    surfaces send hex. Disambiguated by alphabet and length (a 32-byte root is
    64 hex chars vs ~44 base64 chars with ``+/=``), so both encodings
    round-trip correctly.
    """
    if value == "":
        return b""
    is_hex = len(value) % 2 == 0 and bool(_HEX_FULL_RE.match(value))
    looks_base64 = bool(_BASE64_HINT_RE.search(value))
    if is_hex and not looks_base64:
        return hex_to_bytes(value)
    return base64_to_bytes(value)


__all__ = [
    "bytes_to_hex",
    "hex_to_bytes",
    "to_bytes",
    "base64_to_bytes",
    "bytes_to_base64",
    "decode_wire_bytes",
]
